File: scripts/01_load_clean.R
# scripts/01_load_clean.R
# Load and clean 2025 TPE tomato biochar pot experiment data.
# Reads data/2025TPE_raw.xlsx (sheet "summary"), normalizes column names,
# coerces numerics, sets factor levels, and writes data/tpe25.rds.

suppressPackageStartupMessages({
  library(readxl)
  library(dplyr)
  library(tibble)
  library(stringr)
})

set.seed(2025)

raw_path <- "data/2025TPE_raw.xlsx"
out_path <- "data/tpe25.rds"

# Row 1 = column names, row 2 = units, data starts row 3.
# Read names from row 1; read data starting from row 3 with those names.
header <- as.character(unlist(read_excel(raw_path, sheet = "summary",
                                         n_max = 1, col_names = FALSE)))
units_row <- as.character(unlist(read_excel(raw_path, sheet = "summary",
                                            skip = 1, n_max = 1,
                                            col_names = FALSE)))
df <- read_excel(raw_path, sheet = "summary", skip = 2, col_names = FALSE,
                 na = c("", "NA", "#VALUE!", "#DIV/0!", "#N/A"))

# Normalize header: trim spaces, replace problematic chars, drop trailing "*".
clean_name <- function(x) {
  x <- str_trim(x)
  x <- str_replace_all(x, "\\*$", "")
  x <- str_replace_all(x, "/", "_")
  x <- str_replace_all(x, "\\s+", "_")
  x
}
names(df) <- clean_name(header)

# Sanity check: expected 42 columns.
stopifnot(ncol(df) == length(header))

# Coerce numerics for all measurement columns.
id_cols <- c("Treatment", "Rep", "Soil", "Biochar", "Fertilizer")
num_cols <- setdiff(names(df), id_cols)
df <- df %>%
  mutate(across(all_of(num_cols), ~ suppressWarnings(as.numeric(.x))))

# Factor levels per CLAUDE.md.
df <- df %>%
  mutate(
    Treatment  = factor(str_trim(Treatment),
                        levels = paste0("T", 1:8)),
    Rep        = factor(as.integer(Rep), levels = 1:4),
    Soil       = factor(Soil,       levels = c("Red_soil", "Alluvial_soil")),
    Biochar    = factor(Biochar,    levels = c("NO", "Rice_husk", "Leucaena")),
    Fertilizer = factor(Fertilizer, levels = c("NO", "YES"))
  )

# Quick integrity report.
cat("Rows:", nrow(df), " Cols:", ncol(df), "\n")
cat("Treatment x Rep counts (should be 4 reps for each of 8 treatments):\n")
print(addmargins(table(df$Treatment, df$Rep)))
cat("\nNA counts (top 15 columns by missingness):\n")
na_summary <- sort(colSums(is.na(df)), decreasing = TRUE)
print(head(na_summary, 15))

# Persist a units lookup alongside the cleaned data.
units <- tibble(column = names(df), unit = units_row) %>%
  mutate(unit = if_else(column %in% id_cols, NA_character_, unit))

saveRDS(list(data = df, units = units), out_path)
cat("\nWrote", out_path, "\n")

File: scripts/01_load_clean.py
"""01_load_clean.py — load and clean 2025 TPE pot experiment data.

Reads data/2025TPE_raw.xlsx (sheet "summary"), normalizes column names,
coerces numerics (treating "#VALUE!" as NaN), sets categorical levels for the
identifier columns, and writes data/tpe25.pkl as {"data": df, "units": df}.
"""

from __future__ import annotations

import re
from pathlib import Path

import numpy as np
import pandas as pd

RAW_PATH = Path("data/2025TPE_raw.xlsx")
OUT_PATH = Path("data/tpe25.pkl")

ID_COLS = ["Treatment", "Rep", "Soil", "Biochar", "Fertilizer"]
TREATMENT_LEVELS = [f"T{i}" for i in range(1, 9)]
SOIL_LEVELS = ["Red_soil", "Alluvial_soil"]
BIOCHAR_LEVELS = ["NO", "Rice_husk", "Leucaena"]
FERT_LEVELS = ["NO", "YES"]
NA_TOKENS = ["", "NA", "#VALUE!", "#DIV/0!", "#N/A"]


def clean_name(name: str) -> str:
    name = str(name).strip()
    name = re.sub(r"\*$", "", name)
    name = name.replace("/", "_")
    name = re.sub(r"\s+", "_", name)
    return name


def main() -> None:
    np.random.seed(2025)

    # Row 1 = headers, row 2 = units, data starts row 3.
    headers_raw = pd.read_excel(
        RAW_PATH, sheet_name="summary", header=None, nrows=1
    ).iloc[0].tolist()
    units_row = pd.read_excel(
        RAW_PATH, sheet_name="summary", header=None, skiprows=1, nrows=1
    ).iloc[0].tolist()

    df = pd.read_excel(
        RAW_PATH, sheet_name="summary", header=0, skiprows=[1],
        na_values=NA_TOKENS, keep_default_na=True,
    )

    headers = [clean_name(h) for h in headers_raw]
    assert df.shape[1] == len(headers), "Column count mismatch"
    df.columns = headers

    # Coerce all measurement columns to numeric.
    num_cols = [c for c in df.columns if c not in ID_COLS]
    for c in num_cols:
        df[c] = pd.to_numeric(df[c], errors="coerce")

    # Trim string ID cols (preserves NaN via the nullable "string" dtype).
    for c in ("Treatment", "Soil", "Biochar", "Fertilizer"):
        df[c] = df[c].astype("string").str.strip()

    # Drop trailing summary / junk rows: keep only rows with Treatment in T1..T8.
    df = df[df["Treatment"].isin(TREATMENT_LEVELS)].reset_index(drop=True)
    rep_int = pd.to_numeric(df["Rep"], errors="coerce").astype("Int64")

    df["Treatment"] = pd.Categorical(df["Treatment"], categories=TREATMENT_LEVELS, ordered=True)
    df["Soil"] = pd.Categorical(df["Soil"], categories=SOIL_LEVELS)
    df["Biochar"] = pd.Categorical(df["Biochar"], categories=BIOCHAR_LEVELS)
    df["Fertilizer"] = pd.Categorical(df["Fertilizer"], categories=FERT_LEVELS)
    df["Rep"] = pd.Categorical(rep_int, categories=[1, 2, 3, 4], ordered=True)

    # Integrity report.
    print(f"Rows: {len(df)}  Cols: {df.shape[1]}")
    print("\nTreatment x Rep counts (4 reps x 8 treatments expected):")
    print(pd.crosstab(df["Treatment"], df["Rep"], margins=True))
    print("\nTop 15 columns by NA count:")
    print(df.isna().sum().sort_values(ascending=False).head(15))

    # Units lookup (id columns get None).
    units = pd.DataFrame({
        "column": headers,
        "unit": [str(u).strip() if pd.notna(u) else None for u in units_row],
    })
    units.loc[units["column"].isin(ID_COLS), "unit"] = None

    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    pd.to_pickle({"data": df, "units": units}, OUT_PATH)
    print(f"\nWrote {OUT_PATH}")


if __name__ == "__main__":
    main()
